### PR TITLE
Minimum Triangulation Angle

### PR DESCRIPTION
**Pull request by travisdriver**
Rebases #641.

### CODE REVIEW

**Review by ayushbaid — COMMENTED** [review]
Any metrics for this change?

**Comment by travisdriver** [comment]
> Any metrics for this change?

Makes everything much much worse, but this is just on the small datasets. I suspect it will improve performance for the large datasets.

**Comment by travisdriver** [comment]
Even with a 1 degree threshold there are some significant regressions... I'll set the threshold to 0 for now and do more testing on the larger datasets once we merge #643. 

newplot (7)

[visual_comparison_dashboard.html.zip](https://github.com/borglab/gtsfm/files/11795590/visual_comparison_dashboard.html.zip)

File: gtsfm/data_association/point3d_initializer.py [hunk 1]
```diff
@@ -67,8 +69,9 @@ class TriangulationOptions(NamedTuple):
     See the following slides for a derivation of the #req'd samples: http://www.cse.psu.edu/~rtc12/CSE486/lecture15.pdf
 
     Args:
-        reproj_error_threshold: the maximum reprojection error allowed.
         mode: triangulation mode, which dictates whether or not to use robust estimation.
+        reproj_error_threshold: the maximum reprojection error allowed.
+        min_triangulation_angle: the minimum angle (in degrees) subtended at the triangulated point from 2 cameras.
```
Comment (akshay-krishnan) [inline gtsfm/data_association/point3d_initializer.py:74]: mention that this is a threshold, and that tracks for which the max angle at any two cameras < threshold will be rejected.

File: gtsfm/utils/tracks.py [hunk 2]
```diff
@@ -74,3 +77,28 @@ def classify_tracks3d_with_gt_cameras(
         tracks_2d.append(SfmTrack2d(measurements))
 
     return classify_tracks2d_with_gt_cameras(tracks_2d, cameras_gt, reproj_error_thresh_px)
+
+
+def get_triangulation_angle(track3d: SfmTrack, cameras: Dict[int, gtsfm_types.CAMERA_TYPE]) -> float:
```
Comment (akshay-krishnan) [inline gtsfm/utils/tracks.py:82]: I think we want the `max` reflected in the name somehow `get_max_triangulation_angle` ?

File: gtsfm/utils/tracks.py [hunk 3]
```diff
@@ -74,3 +77,28 @@ def classify_tracks3d_with_gt_cameras(
...
+    """Get the angle subtended by the cameras at the 3D landmark of the track.
```
Comment (akshay-krishnan) [inline gtsfm/utils/tracks.py:83]: add "in degrees", ideally might be good to have that in the name as well.

File: gtsfm/data_association/point3d_initializer.py [hunk 4]
```diff
@@ -41,6 +42,7 @@ class TriangulationExitCode(Enum):
     INLIERS_UNDERCONSTRAINED = 2  # insufficent number of inlier measurements
     POSES_UNDERCONSTRAINED = 3  # insufficent number of estimated camera poses
     EXCEEDS_REPROJ_THRESH = 4  # estimated 3d point exceeds reprojection threshold
+    LOW_TRIANGULATION_ANGLE = 5  # estimated 3d point exceeds reprojection threshold
```
Comment (akshay-krishnan) [inline gtsfm/data_association/point3d_initializer.py:45]: update the comment after copying it? :)

**Review by akshay-krishnan — APPROVED** [review]

File: gtsfm/data_association/point3d_initializer.py (hunk 4 above)
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:45]: Oopsies

**Comment by akshay-krishnan** [comment]
> Even with a 1 degree threshold there are some significant regressions... I'll set the threshold to 0 for now and do more testing on the larger datasets once we merge #643.

It might help to add a "triangulation_angle" metric to point3d initializer, just like we have track length. I suspect this might be removing a lot of the length 2 tracks that we found to be useful?

File: gtsfm/utils/tracks.py [hunk 5]
```diff
@@ -74,3 +77,28 @@ def classify_tracks3d_with_gt_cameras(
...
+    camera_ind: List[int] = []
+    for k in range(track3d.numberMeasurements()):
+        i, _ = track3d.measurement(k)
+        camera_ind.append(i)
+
+    angles: List[float] = []
+    for i1, i2 in itertools.combinations(camera_ind, 2):
+        ray1 = track3d.point3() - cameras[i1].pose().translation()
+        ray2 = track3d.point3() - cameras[i2].pose().translation()
+        angles.append(np.rad2deg(np.arccos(np.dot(ray1, ray2) / (np.linalg.norm(ray1) * np.linalg.norm(ray2)))))
```
Comment (johnwlambert) [inline gtsfm/utils/tracks.py:102]: I believe we have an existing function that shares similar functionality -- https://github.com/borglab/gtsfm/blob/bf58ea2d48711d7047ffb195fa053a1b81fe270f/gtsfm/densify/mvs_utils.py#L53

Could we re-use that?

**Pull request merged**